### DISCUSSION TITLE
"Synthetic" gem constraints

### DISCUSSION THREAD

**Discussion by camertron** [ruby/rubygems#5383] [Ideas]
### Describe the problem as clearly as you can

This isn't a bug or a problem per se. I'm hoping to have a discussion about a feature I've been thinking about, but I haven't yet convinced myself it's a good idea.

For a while now, bundler has shipped with a feature that lets you specify the version(s) of Ruby your project supports. I seem to recall reading somewhere that the Ruby version constraint is implemented as if it were another gem whose version must agree with all the other gems in the bundle. What if such a synthetic version constraint could generalized and exposed as a public API?

Specifically I'm thinking of being able to specify certain versions of 3rd-party APIs a library supports. For example, I'm working on a tool that makes it easier to deploy Rails apps into Kubernetes clusters, and it would be convenient to be able to specify the Kubernetes version along with the gem dependencies. That way, the bundle would fail to resolve if a particular gem didn't meet the minimum Kubernetes version requirements.

Something like this, maybe:

```ruby
source 'https://rubygems.org'

ruby '3.1.0'
constraint(:kubernetes, '>= 1.19', '< 1.23')

gem 'foo'
# etc etc
```

I'm curious to hear what others think about this, and whether or not it's a terrible idea.

One could emulate the desired behavior by publishing an empty Kubernetes gem, one version for each released Kubernetes version. That feels much worse.

**Comment by simi** [comment]
:wave: Hello. Thanks for sharing your idea. I have moved this to discussions since it is not (as stated in description issue).

From my point of view RubyGems/Bundler is there to install RubyGems based on their dependencies. I don't think anything non-related should be part of the resolution logic. You can indeed create "meta" gem. Also you can use gemspec metadata to store custom additional info, but it will not affect dependencies resolution.

**Comment by hsbt** [comment]
This cannot work from the Gemfile alone. The Ruby constraint bites because every gem declares `required_ruby_version` and Bundler turns that declaration into a dependency on a synthetic `Ruby` package, so a Gemfile constraint and a gem's own requirement meet on the same name. Nothing declares a Kubernetes requirement, so `constraint(:kubernetes, ">= 1.19")` would have no counterpart to conflict with and no resolution could ever fail on it.

**Comment by camertron** [comment]
Hi @hsbt, thanks for your response 😄 

Reading back over my initial post, I didn't do a great job describing what I had in mind. I understand that Rubygems can't (and shouldn't) know what version of eg. Kubernetes you have installed and therefore has no way of determining a requirement that the constraints can be matched against. However that's not really what I meant.

My original idea was to fail bundle resolution if two gems declare constraints that disagree with each other. For example:

```ruby
# gem 1
constraint :kubernetes, '>= 1.19', '< 1.23'

# gem 2
constraint :kubernetes, '> 1.23'
```

There's no requirement to compare, but these two constraints cannot be mutually satisfied, so bundle resolution fails.

@simi, apologies for missing your comment back in 2022 😬 

> From my point of view RubyGems/Bundler is there to install RubyGems based on their dependencies. I don't think anything non-related should be part of the resolution logic.

Sure, that's a totally reasonable critique of my idea. And of course there are other ways to announce that a gem is or is not compatible with some external thing (raise an error, include supported versions in README, etc). I would say that one of bundler's "jobs" is to tell me if a gem I'm adding is incompatible with other gems in my bundle. Incompatibility can take many forms, and I think it's reasonable to consider external compatibility as well as inter-gem compatibility.

> You can indeed create "meta" gem.

That doesn't really help me because the constraints I want to declare are not gem constraints.

> Also you can use gemspec metadata to store custom additional info, but it will not affect dependencies resolution.

That's an interesting idea! Perhaps some metadata plus a bundler plugin could get me 90% of the way there 😄
